---
title: "2021-07-13_public-data-google-sheet-day1"
author: "Jane Kim"
date: "7/13/2021"
output: html_document
---

Google Sheet with the list of scATAC-seq data:

https://docs.google.com/spreadsheets/d/1IJhYMysjt-gyFeG1P7BL0NbKkOl1QPKqljz7s6SLZ6U/edit#gid=1161067356

1. Data Location: GSE139369
  1) Important features 
    - 10X genomics
    - Human

2. Google GSE139369: https://www.google.com/search?q=GSE139369&oq=GSE139369&aqs=chrome..69i57.347j0j7&sourceid=chrome&ie=UTF-8
  1) click the very top search result
  2) land on this page: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE139369
  
3. Scroll down to find "Relations". Copy the serial number next to "BioProject" which is PRJNA579391.

4. Paste and search the serial number at https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA579391&o=acc_s%3Aa

5. Obtain all the SRR 

Commands

$ MAESTRO scatac-init --batch --bulk_peaks --downsample \
--target_reads 5000000 --input_path /liulab/galib/sc_CIDC/MAESTRO/test/tutorial/data/time_data \
--gzip --platform 10x-genomics --format fastq --species GRCh38 \
--deduplication cell-level --mapping chromap \
--giggleannotation /liulab/galib/sc_CIDC/annotations/giggle.all/ \
--fasta /liulab/galib/sc_CIDC/references/Refdata_scATAC_MAESTRO_GRCh38_1.1.0/GRCh38_genome.fa \
--index /liulab/galib/sc_CIDC/references/chromap/GRCh38_chromap.index \
--whitelist /liulab/galib/sc_CIDC/references/whitelist/737K-cratac-v1.txt \
--cores 16 --directory multi-scatac-chromap \
--annotation --method RP-based --signature human.immune.CIBERSORT \
--clusterpeak --shortpeak \
--rpmodel Enhanced \
--peak_cutoff 100 --count_cutoff 1000 --frip_cutoff 0.2 --cell_cutoff 500

```txt
--batch TRUE_OR_FALSE

when no text follows it, set as "True". 
True means that all the samples are from the same experiment. If so, peaks from all samples will be merged and a count matrix for each sample will be produced based on the merged peak set

not explicitly mentioned, but when --batch is "True", --consensus_peaks and --cutoff_samples are set to False and 2, respectively.

if batch processing is True, when merging peaks from all samples and the user want to call peaks from each sample individually and merge the peaks to get consensus peaks , only keep the peaks present in at least cutoff_samples samples

-consensus_peaks TRUE_OR_FALSE

--cutoff-samples NUMBER

Minimum number of samples to present consensus peaks. The peaks present in at least "cutoff_samples" will be kept

```

```txt
--bulk_peaks TRUE_OR_FALSE

For multi-samples from the same experiment, if set as true, peaks will be called after merging all bam file. Bulk_peaks and consensus_peaks are mutually exclusive.

```

consensus + cutoff  VS bulk is mutually exclusive, the pipeline will only go through one of them. The default is the latter one: calling bulk peaks.


```txt
--downsample TRUE_OR_FALSE

For deeply sequenced samples, bam files can be downsampled to a certain number of reads (target_reads) to get peak set. 

```

```txt
--target_reads NUMBER

Number of reads to be kept in downsampling. If the number of reads in a sample is fewer than the target_reads, the original number of reads will be kept. 
```

```txt
--input_path PATH

Path where the fastq files are locatted

```

```txt
--gzip TRUE_OR_FALSE

If the input files are gzipped, set as "True".

```

```txt
--platform {10x-genomics, sci-ATAC-seq, microfluidic}
```

```txt
--format {fastq, fragments, bam}

The format of the input files.

If the platform is 10x-genomics, users can start with sequencing fastq files, bam files with CB tag or fragments.tsv.gz file generated by CellRanger ATAC.
If the platform is sci-ATAC-seq, users can start with sequencing fastq files, bam files with CB tag.
If the platform is microfluidic, users can only start with sequencing fastq files.
If the format is 'fastq', fastqdir and fastqprefix need to be set.
If the format is 'bam', bam needs to be set.
If the format is 'fragments', frag needs to be set.

```

```txt
--species {GRCh38, GRCm38}

GRCh38 for human and CRCm38 for mouse
```

```txt
--deduplication {cell-level, bulk-level}

```

```txt
--mapping {chromap, minimap2}

Choose the alignment tool for scATAC-seq from either chromap or minimap2

```

```txt
--giggleannotation

Path to the giggle annotation file required for regulator identification
```

```txt
--fasta

Genome fastq file for mapping

Human: ./Refdata_scATAC_MAESTRO_GRCh38_1.1.0/GRCh38_genome.fa
Mouse: ./Refdata_scATAC_MAESTRO_GRCh38_1.1.0/GRCm38_genome.fa
```

```txt
--index

Path to the reference index file for chromap.
When using chromap, you need reference index.
```

```txt
--whitelist PATH_TO_FILE

ONLY FOR SCI-ATAC-SEQ AND 10X-GENOMICS

If the platform is 'sci-ATAC-seq' or '10x-genomics', specify the barcode library (whitelist) so that the pipeline can correct cell barcodes with 1 base mismatched. Otherwise, the pipeline will automatically output the barcodes with enough read counts ( >1000).
```

```txt
--cores NUMBER

Number of cores to use
```

```txt
--directory PATH

Path where the workflow is initialized and results will be stored. Default is MAESTRO.
```

```txt
--annotation

If --annotation is set to TRUE, --method should also be specified. Whether or not to perform cell-type annotation. 
```

```txt
--method {RP-based, peak-based, both}

```

```txt
--signature {'human.immune.CIBERSORT', 'mouse.brain.ALLEN', 'mouse.all.facs.TabulaMuris', 'mouse.all.droplet.TabulaMuris'}

Required when --method is set as "RP-based (default)"
Cell signature file used to annotate cell types
A custom cell signature can be provided as a path. The file must be tab-separated without a header, with the first column as "cell type" and the second column as "signature gene".
```

--clusterpeak and --shortpeak are two of three customized peaks. When they are not explicitly called, the default is false, but when they are called it is set to true. 

```txt
--clusterpeak TRUE_OR_FALSE

Whether or not to call peaks by cluster
```

```txt
--shortpeak TRUE_OR_FALSE

If it's set True, MAESTRO will split the fragment file according to clustering result, and then call peaks for each cluster.

```

```txt
--rpmodel {Simple, Enhanced}

For each gene, simple model sums over the impact of all regulatory elements within the up/dowm-stream of TSS.
On the basis of simple model, Enhanced model gives the regulatory elements within the exon region a higher weight, and also excludes the regulatory elements overlapped with another gene (the promoter and exon of a nearby gene).
```

```txt
--peak_cutoff NUMBER

Minimum number of peaks included in each cell.
```

```txt
--count_cutoff NUMBER

Cutoff for the number of counts in each cell
```

```txt
--frip_cutoff NUMBER

Cutoff for fraction of reads in promotor in each cell
```

```txt
--cell_cutoff NUMBER

Minimum number of cells covered by each peak
```